### PR TITLE
fix: [CDS-78304]: Check session connectivity before cleanup

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
While cleaning to files if the session gets disconnected clean might fail, to handle this we will create a new session check if the session had lost connectivity to do the clean up.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)

### CODE REVIEW

**Comment by yogesh-chauhan** [comment]
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java` line 77 at r3](https://reviewable.io/reviews/harness/harness-core/53143#-NewuvzNEkl5TVbmdrnj:-NewuvzNEkl5TVbmdrnk:b-nsug4s) ([raw file](https://github.com/harness/harness-core/blob/2b38136adc909b9d4182b0635ed2ef1cfae956b3/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java#L77)):*
> ```Java
>   private final LogCallback logCallback;
> 
>   WinRmClient client;
> ```

instead of this, we can create a new method "check Connection "

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java [hunk 1]
```diff
@@ -80,6 +80,10 @@ public class WinRmSession implements AutoCloseable {
   private Path cacheFilePath;
   private final AuthenticationScheme authenticationScheme;
 
+  public void checkConnectivity() {
+    client.createShell();
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java:84]: is this covering kerberos also?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java [hunk 2]
```diff
@@ -228,6 +229,26 @@ public static void cleanupFiles(WinRmSession session, String file, String powers
     }
   }
 
+  public static void cleanupFiles(WinRmSession session, String file, String powershell, boolean disableCommandEncoding,
```
Comment (acasian-harness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java:232]: imho all clean up activity should be done part of clean command unit.

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java` line 232 at r4](https://reviewable.io/reviews/harness/harness-core/53143#gh-1334576815:-Nexv2OnCrMQU2MSFlWd:bcc1cnh) ([raw file](https://github.com/harness/harness-core/blob/96315624339c51e66a200986c4339874db811e5e/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java#L232)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

imho all clean up activity should be done part of clean command unit.
</blockquote></details>

With the customer use case if the pipeline is aborted clean command unit will not run hence the files won't be cleaned up, also during the execution we create the file with name harness-executable-{executionId} for each command unit.
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java` line 77 at r3](https://reviewable.io/reviews/harness/harness-core/53143#-NewuvzNEkl5TVbmdrnj:-Nexv_AzD5lgJImS2v5w:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/2b38136adc909b9d4182b0635ed2ef1cfae956b3/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java#L77)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

instead of this, we can create a new method "check Connection "
</blockquote></details>

Done
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java` line 84 at r4](https://reviewable.io/reviews/harness/harness-core/53143#gh-1334576727:-NexvaAt6SGfAIYYZfCj:b2nhswy) ([raw file](https://github.com/harness/harness-core/blob/96315624339c51e66a200986c4339874db811e5e/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java#L84)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

is this covering kerberos also?
</blockquote></details>

Yes

**Comment by happyhebaby** [comment]
Could you please remove me from the thread as my mailbox is flooded with the discussion?
    On Friday, September 22, 2023 at 10:29:59 AM PDT, Sarthak Kasat ***@***.***> wrote:  
 
 


@sarthakkasat commented on this pull request.


Reviewable status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained.


930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java line 232 at r4 (raw file):
Previously, acasian-harness (Alexandru Casian) wrote…

imho all clean up activity should be done part of clean command unit.

 
With the customer use case if the pipeline is aborted clean command unit will not run hence the files won't be cleaned up, also during the execution we create the file with name harness-executable-{executionId} for each command unit.

930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java line 77 at r3 (raw file):
Previously, yogesh-chauhan (Yogesh Chauhan) wrote…

instead of this, we can create a new method "check Connection "

 
Done

930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java line 84 at r4 (raw file):
Previously, acasian-harness (Alexandru Casian) wrote…

is this covering kerberos also?

 
Yes

**Review by ivan-83 — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 1 of 1 files at r2, 1 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java` line 232 at r4](https://reviewable.io/reviews/harness/harness-core/53143#gh-1334576815:-Nf7aHWy1dKNl3vZ1hMa:bh5tqco) ([raw file](https://github.com/harness/harness-core/blob/96315624339c51e66a200986c4339874db811e5e/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java#L232)):*
<details><summary><i>Previously, sarthakkasat (Sarthak Kasat) wrote…</i></summary><blockquote>

With the customer use case if the pipeline is aborted clean command unit will not run hence the files won't be cleaned up, also during the execution we create the file with name harness-executable-{executionId} for each command unit.
</blockquote></details>

I think we need cleanup action and for SSH and WinRm, the solution could be to repopulate the command task params on the abort pipeline, send params from the manager to delegate, and call SshCleanupCommandHandler or WinRmCleanupCommandHandler. 

QQ, The cleanupFiles method should be called after commands execution because it's already part of finally block. Why it's not called for abort pipeline flow?

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)* status: 4 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java` line 232 at r4](https://reviewable.io/reviews/harness/harness-core/53143#gh-1334576815:-Nf9nGV7DNdv5N3ovk7S:b-95wp68) ([raw file](https://github.com/harness/harness-core/blob/96315624339c51e66a200986c4339874db811e5e/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmExecutorHelper.java#L232)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

I think we need cleanup action and for SSH and WinRm, the solution could be to repopulate the command task params on the abort pipeline, send params from the manager to delegate, and call SshCleanupCommandHandler or WinRmCleanupCommandHandler. 

QQ, The cleanupFiles method should be called after commands execution because it's already part of finally block. Why it's not called for abort pipeline flow?
</blockquote></details>

The clean up method is called in finally block, this was the change that we had merged recently as a part of same ticket. Before that the clean up logic was not in finally block.

**Comment by sarthakkasat** [comment]
trigger smartchecks

**Review by acasian-harness — APPROVED** [review]

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)* status: 3 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by sarthakkasat** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53143)* status: 2 file(s) to review. 2. unresolved discussions. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java` line 84 at r4](https://reviewable.io/reviews/harness/harness-core/53143#gh-1334576727:-NfGF12C11bqTN4Nz-SE:bhie5u6) ([raw file](https://github.com/harness/harness-core/blob/96315624339c51e66a200986c4339874db811e5e/930-delegate-tasks/src/main/java/io/harness/delegate/task/winrm/WinRmSession.java#L84)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

I think it will not cover kerberos. client would be null in that case
</blockquote></details>

As discussed, lets handle in next PR

**Comment by sarthakkasat** [comment]
trigger review

**Comment by sarthakkasat** [comment]
trigger review

**Pull request merged**